### PR TITLE
[DEV-2591] Mass Cancel

### PR DESCRIPTION
**Pull request by mihaiberq**
Extend platforms to allow for sale cancelation. Decided to expose it on the `nft` rather than `offer`.

### CODE REVIEW

File: src/refinable/contract/ContractWrapper.ts [hunk 1]
```diff
@@ -119,6 +120,10 @@ export class ContractWrapper implements IContractWrapper {
       callOverrides
     );
 
+    if (typeof callback === "function") {
+      callback();
```
Comment (sneljo1) [inline src/refinable/contract/ContractWrapper.ts:124]: Can we make this less generic?

File: src/platform/OpenseaPlatform.ts [hunk 2]
```diff
@@ -315,6 +321,77 @@ export class OpenseaPlatform extends AbstractPlatform {
...
+    });
+
+    const builder = new Seaport.Builders.SingleToken(this.refinable.chainId);
+
+    const { orderParams: { parameters, signature } = {} } = offer;
+
+    const nonce = await this.getNonce(this.refinable.accountAddress);
+
+    // THIS HAS TO BE A 1:1 MATCH TO order.orderParams
+    const builtOrder = builder.build({
```
Comment (sneljo1) [inline src/platform/OpenseaPlatform.ts:354]: Cant we just use `orderParams`?

File: src/nft/ERC1155NFT.ts [hunk 3]
```diff
@@ -160,6 +160,16 @@ export class ERC1155NFT extends AbstractEvmNFT {
     );
   }
 
+  async cancelSaleOffers({
```
Comment (sneljo1) [inline src/nft/ERC1155NFT.ts:163]: So, it only works with 721?

File: src/nft/ERC721NFT.ts [hunk 4]
```diff
@@ -345,6 +350,94 @@ export class ERC721NFT extends AbstractEvmNFT {
     );
   }
 
+  async cancelSaleOffers({
+    offers,
+    confirmations,
```
Comment (sneljo1) [inline src/nft/ERC721NFT.ts:355]: Not sure if the confirmations are really needed. Doesnt seem used

**Review by sneljo1 — APPROVED** [review]

File: src/platform/OpenseaPlatform.ts (hunk 2 above)
Comment (mihaiberq) [reply src/platform/OpenseaPlatform.ts:354]: sadly, it's more difficult. there are fields that are ignored when doing `builtOffer.params`.

File: src/nft/ERC1155NFT.ts (hunk 3 above)
Comment (mihaiberq) [reply src/nft/ERC1155NFT.ts:163]: our list for sale on other marketplaces/buying on other marketplaces works only for 721s

File: src/refinable/contract/ContractWrapper.ts (hunk 1 above)
Comment (mihaiberq) [reply src/refinable/contract/ContractWrapper.ts:124]: can do

File: src/nft/ERC1155NFT.ts (hunk 3 above)
Comment (sneljo1) [reply src/nft/ERC1155NFT.ts:168]: > our list for sale on other marketplaces/buying on other marketplaces works only for 721s
But we index other external listings right

**Pull request merged**